Correct the Rudder.pm inventory, to be compatible with Windows and add the server-roles if necessary

--- fusioninventory-agent/lib/FusionInventory/Agent/Inventory.pm	2016-08-30 18:14:54.524244071 +0200
+++ fusioninventory-agent/lib/FusionInventory/Agent/Inventory.pm	2016-08-30 18:15:06.772291754 +0200
@@ -68,7 +68,7 @@
     PROCESSES        => [ qw/USER PID CPUUSAGE MEM VIRTUALMEMORY TTY STARTED
                              CMD/ ],
     REGISTRY         => [ qw/NAME REGVALUE HIVE/ ],
-    RUDDER           => [ qw/AGENT UUID HOSTNAME/ ],
+    RUDDER           => [ qw/AGENT UUID HOSTNAME SERVER_ROLE/ ],
     SLOTS            => [ qw/DESCRIPTION DESIGNATION NAME STATUS/ ],
     SOFTWARES        => [ qw/COMMENTS FILESIZE FOLDER FROM HELPLINK INSTALLDATE
                             NAME NO_REMOVE RELEASE_TYPE PUBLISHER

--- fusioninventory-agent/lib/FusionInventory/Agent/Task/Inventory/Generic/Rudder.pm	2016-08-30 18:05:06.181949600 +0200
+++ fusioninventory-agent/lib/FusionInventory/Agent/Task/Inventory/Generic/Rudder.pm	2016-08-30 18:13:54.848011697 +0200
@@ -7,8 +7,10 @@
 
 use FusionInventory::Agent::Tools;
 
+my $uuid_hive = ($OSNAME eq 'MSWin32') ? 'C:\Program Files\Rudder\etc\uuid.hive' : '/opt/rudder/etc/uuid.hive';
+
 sub isEnabled {
-    return -r '/opt/rudder/etc/uuid.hive';
+    return -r $uuid_hive;
 }
 
 sub doInventory {
@@ -17,23 +19,28 @@
     my $inventory = $params{inventory};
     my $logger    = $params{logger};
 
-    # get Rudder UUID
-    my $Uuid = getFirstLine(
-        logger => $logger, file => '/opt/rudder/etc/uuid.hive'
+    # Get the Rudder UUID
+    my $uuid = getFirstLine(
+        logger => $logger, file => "$uuid_hive"
     );
-    # get all agents running on that machine
+    # Get all agents running on the local machine
     my @agents = _manageAgent(
-        logger => $logger, command => 'ls /var/rudder'
+        logger => $logger
     );
-    # get machine hostname
+    # Get the machine hostname
     my $command = $OSNAME eq 'linux' ? 'hostname --fqdn' : 'hostname';
     my $hostname = getFirstLine(
         logger => $logger, command => $command
     );
+    # Get theserver roles
+    my @serverRoles = _listServerRoles();
+
+    # Output all of those in <RUDDER>
     my $rudder = {
         HOSTNAME => $hostname,
-        UUID => $Uuid,
+        UUID => $uuid,
         AGENT => \@agents,
+        SERVER_ROLE => \@serverRoles,
     };
 
     $inventory->addEntry(
@@ -41,23 +48,45 @@
     );
 }
 
+sub _listServerRoles {
+    my $server_roles_dir = ($OSNAME eq 'MSWin32') ? 'C:\Program Files\Rudder\etc\server-roles.d' : '/opt/rudder/etc/server-roles.d';
+    my @server_roles;
+
+    if (-d "$server_roles_dir") {
+        opendir(DIR, $server_roles_dir);
+
+        while (my $file = readdir(DIR)) {
+            # Use a regular expression to ignore files beginning with a period
+            next if ($file =~ m/^\./);
+            push @server_roles, $file;
+        }
+        closedir(DIR);
+    }
+
+    return @server_roles;
+}
 sub _manageAgent {
-    my $handle = getFileHandle(@_);
     my %params = @_;
     my $logger = $params{logger};
 
     my @agents;
 
-    # each line could be a new agent
-    while(my $name = <$handle>){
-
-        chomp $name;
-        # verify agent name
-        next unless $name =~ /cfengine/;
-
-        my $server_hostname_file = "/var/rudder/$name/policy_server.dat";
-        my $uuid_file            = "/var/rudder/$name/rudder-server-uuid.txt";
-        my $cfengine_key_file    = "/var/rudder/$name/ppkeys/localhost.pub";
+    # Potential agent directory candidates
+    my %agent_candidates = ( '/var/rudder/cfengine-community' => 'cfengine-community',
+                             '/var/cfengine'                  => 'cfengine-nova',
+                             'C:/Program Files/Cfengine'      => 'cfengine-nova',
+                            );
+
+    foreach my $candidate (keys(%agent_candidates)){
+
+        # Verify if the candidate is installed and configured
+        next unless ( -e "${candidate}/policy_server.dat" );
+
+        # Get a list of useful file paths to key Rudder components
+        my $agent_name           = "$agent_candidates{${candidate}}";
+        my $server_hostname_file = "${candidate}/policy_server.dat";
+        my $uuid_file            = "${candidate}/rudder-server-uuid.txt";
+        my $cfengine_key_file    = "${candidate}/ppkeys/localhost.pub";
 
         # get policy server hostname
         my $serverHostname = getFirstLine (
@@ -66,32 +95,32 @@
         );
         chomp $serverHostname;
 
-        # get policy server uuid
+        # Get the policy server UUID
         #
-        # the default file is no longer /var/rudder/tmp/uuid.txt since the
+        # The default file is no longer /var/rudder/tmp/uuid.txt since the
         # change in http://www.rudder-project.org/redmine/issues/2443.
-        # we gracefully fallback to the old default if we can not find the
-        # new file.
+        # We gracefully fallback to the old default if the new file cannot
+        # be found.
         my $serverUuid = getFirstLine (
             logger => $logger,
             file => ( -e "$uuid_file" ) ? $uuid_file : "/var/rudder/tmp/uuid.txt"
         );
         chomp $serverUuid;
 
-        # get CFengine public key
+        # Get the CFengine public key
         my $cfengineKey = getAllLines(
             file => $cfengine_key_file
         );
 
-        # get owner name
+        # Get the current user name
         my $owner = getFirstLine (
             logger => $logger,
             command => 'whoami'
         );
 
-        # build agent from datas
+        # List the current agent facts
         my $agent = {
-            AGENT_NAME             => $name,
+            AGENT_NAME             => $agent_name,
             POLICY_SERVER_HOSTNAME => $serverHostname,
             CFENGINE_KEY           => $cfengineKey,
             OWNER                  => $owner,
@@ -102,7 +131,6 @@
 
     }
 
-    close $handle;
     return @agents;
 }
